SU2_CFD/src/solver_direct_mean.cpp

```diff
@@ -12483,13 +12483,241 @@ void CEulerSolver::BC_Engine_Exhaust(CGeometry *geometry, CSolver **solver_conta
+      /*---         but the primitive gradients of the reflected state have to be   ---*/
+      /*---         determined additionally such that symmetry at the boundary is   ---*/
+      /*---         enforced. Based on the Viscous_Residual routine.                ---*/
+      /*-------------------------------------------------------------------------------*/
+      if (config->GetViscous()) {
+        
+        /*--- Set the normal vector and the coordinates. ---*/
+        visc_numerics->SetNormal(Normal);
+        visc_numerics->SetCoord(geometry->node[iPoint]->GetCoord(), geometry->node[iPoint]->GetCoord());
+
+        /*--- Set the primitive and Secondary variables. ---*/
+        visc_numerics->SetPrimitive(V_domain, V_reflected);
+        visc_numerics->SetSecondary(node[iPoint]->GetSecondary(), node[iPoint]->GetSecondary());
+        
+        /*--- For viscous Fluxes also the gradients of the primitives need to be determined.
+              1. The gradients of scalars are mirrored along the sym plane just as velocity for the primitives
+              2. The gradients of the velocity components need more attention, i.e. the gradient of the
+                 normal velocity in tangential direction is mirrored and the gradient of the tangential velocity in 
+                 normal direction is mirrored. ---*/
+
+        /*--- Get gradients of primitives of boundary cell ---*/
+        for (iVar = 0; iVar < nPrimVarGrad; iVar++)
+          for (iDim = 0; iDim < nDim; iDim++)
+            Grad_Prim[iVar][iDim] = node[iPoint]->GetGradient_Primitive(iVar, iDim);
+
+        /*--- Reflect the gradients for all scalars including the velocity components.
+              The gradients of the velocity components are overriden later with the 
+              correct values: grad(V)_r = grad(V) - 2 [grad(V)*n]n, V beeing any primitive ---*/
+        for (iVar = 0; iVar < nPrimVarGrad; iVar++) {
+
+          /*--- Compute projected part of the gradient in a dot product ---*/
+          ProjGradient = 0.0;
+          for (iDim = 0; iDim < nDim; iDim++)
+            ProjGradient += Grad_Prim[iVar][iDim]*UnitNormal[iDim];
+
+          for (iDim = 0; iDim < nDim; iDim++)
+            Grad_Reflected[iVar][iDim] = Grad_Prim[iVar][iDim] - 2.0 * ProjGradient*UnitNormal[iDim];
+        }
+        
+        /*--- Compute unit tangential, the direction is arbitrary as long as t*n=0 ---*/
+        switch( nDim ) {
+          case 2: {
+            Tangential[0] = -UnitNormal[1];
+            Tangential[1] =  UnitNormal[0];
+            break;
+          }
+          case 3: {
+            /*--- Find the largest entry index of the UnitNormal, and create Tangential vector based on that. ---*/
+            unsigned short Largest, Arbitrary, Zero;
+            if     (abs(UnitNormal[0]) >= abs(UnitNormal[1]) && abs(UnitNormal[0]) >= abs(UnitNormal[2])){Largest=0;Arbitrary=1;Zero=2;}
```

Comment:
As we should always have a plane for this BC type, is this something you can check outside the loop, or complete as a preprocessing and store? Looks a little expensive to perform at every vertex during runtime

Reply:
Yep, moved that out of the loop. The unit-normal as well.